internal/sriov/sriov.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sriov
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/vishvananda/netlink"
+)
+
+var (
+	sysClassNet = "/sys/class/net/"
+	tcpProtoStr = "tcp"
+	udpProtoStr = "udp"
+)
+
+func IsSriovPF(ifName string) bool {
+	ifPfDir := filepath.Join(sysClassNet, ifName, "device", "physfn")
+	if _, err := os.Stat(ifPfDir); err != nil {
+		return false
+	}
+	return true
+}
+
+func GetSriovVFs(ifName string) ([]string, error) {
+	vfNames := []string{}
+	if !IsSriovPF(ifName) {
+		return nil, fmt.Errorf("interface %s is not a sriov pf type", ifName)
+	}
+	vfFnsDir := filepath.Join(sysClassNet, ifName, "device", "virtfn*")
+	vfsDir, err := filepath.Glob(vfFnsDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, vfDir := range vfsDir {
+		vfDirInfo, err := ioutil.ReadDir(vfDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read vf directory %s: %q", vfDir, err)
+		}
+		vfLink, err := netlink.LinkByName(vfDirInfo[0].Name())
+		if err != nil {
+			vfNames = append(vfNames, vfLink.Attrs().Name)
+		}
+	}
+	return vfNames, nil
+}
+
+func UnprotectPorts(logger log.Logger, ifName string, protocol string, ports []int) error {
+	if _, err := exec.LookPath("tc"); err != nil {
+		return err
+	}
+	for _, port := range ports {
+		cmd := exec.Command("bash", "-c", fmt.Sprintf("tc filter add dev %s ingress pref 9999 protocol ip u32 match %s dst %s 0xffff action ok index 99", ifName, protocol, strconv.Itoa(port)))
```

Comment:
Do we need to invoke bash?  I think we could just run the tc command directly, couldn't we?